🐛 BUG: Crashes on reload if using `node:xxx` imports

### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.28.0

### What version of Node are you using?

20.11.0

### What operating system and version are you using?

Windows 11

### Describe the Bug

In the latest version of wrangler, when you make a change in a worker that imports anything from `node:`, it seems to crash and be unable to rebuild.

`wrangler.toml`:
```toml
name = "wandering-butterfly-5921"
main = "src/index.ts"
compatibility_date = "2024-02-08"
compatibility_flags = ["nodejs_compat"]

```

`src/index.ts`:
```typescript
import { Buffer } from 'node:buffer';

export interface Env {}
export default {
	async fetch(request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
		const str = Buffer.from('Hello World!').toString('base64');
		return new Response(str)
	},
};
```

Steps to reproduce:

- Use worker config and worker above
- npx wrangler@3.28.0 dev
- Make a change to `src/index.ts`
- Observe wrangler crashing with the following:

```
✘ [ERROR] Could not resolve "node:buffer"

    src/index.ts:1:23:
      1 │ import { Buffer } from 'node:buffer';
        ╵                        ~~~~~~~~~~~~~

  The package "node:buffer" wasn't found on the file system but is built into node.
  Add "node_compat = true" to your wrangler.toml file and make sure to prefix the module name with "node:" to enable Node.js compatibility.
```

--- a/.changeset/cuddly-apples-divide.md
+++ b/.changeset/cuddly-apples-divide.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+fix: ensure `wrangler dev` can reload without crashing when importing `node:*` modules
+
+The previous Wrangler release introduced a regression that caused reloads to fail when importing `node:*` modules. This change fixes that, and ensures these modules can always be resolved.

--- a/packages/wrangler/e2e/dev.test.ts
+++ b/packages/wrangler/e2e/dev.test.ts
@@ -162,6 +162,7 @@ describe("basic dev tests", () => {
 					name = "${workerName}"
 					main = "src/index.ts"
 					compatibility_date = "2023-01-01"
+					compatibility_flags = ["nodejs_compat"]
 
 					[vars]
 					KEY = "value"
@@ -195,9 +196,12 @@ describe("basic dev tests", () => {
 
 			await worker.seed({
 				"src/index.ts": dedent`
+						import { Buffer } from "node:buffer";
 						export default {
 							fetch(request, env) {
-								return new Response("Updated Worker! " + env.KEY)
+								const base64Message = Buffer.from("Updated Worker!").toString("base64");
+								const message = Buffer.from(base64Message, "base64").toString();
+								return new Response(message + " " + env.KEY)
 							}
 						}`,
 			});
@@ -216,6 +220,7 @@ describe("basic dev tests", () => {
 						name = "${workerName}"
 						main = "src/index.ts"
 						compatibility_date = "2023-01-01"
+						compatibility_flags = ["nodejs_compat"]
 
 						[vars]
 						KEY = "updated"

--- a/packages/wrangler/src/__tests__/navigator-user-agent.test.ts
+++ b/packages/wrangler/src/__tests__/navigator-user-agent.test.ts
@@ -72,7 +72,7 @@ describe("isNavigatorDefined", () => {
 			assert(false, "Unreachable");
 		} catch (e) {
 			expect(e).toMatchInlineSnapshot(
-				`[AssertionError: Can't both enable and disable a flag]`
+				`[Error: Can't both enable and disable a flag]`
 			);
 		}
 	});

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts
@@ -12,10 +12,10 @@ export const nodejsCompatPlugin: (silenceWarnings: boolean) => Plugin = (
 	name: "nodejs_compat imports plugin",
 	setup(pluginBuild) {
 		// Infinite loop detection
-		const seen = new Set();
+		const seen = new Set<string>();
 
 		// Prevent multiple warnings per package
-		const warnedPackaged = new Map();
+		const warnedPackaged = new Map<string, string[]>();
 
 		pluginBuild.onStart(() => {
 			seen.clear();
@@ -41,14 +41,12 @@ export const nodejsCompatPlugin: (silenceWarnings: boolean) => Plugin = (
 					// esbuild couldn't resolve the package
 					// We should warn the user, but not fail the build
 
-					if (!warnedPackaged.has(path)) {
-						warnedPackaged.set(path, [opts.importer]);
-					} else {
-						warnedPackaged.set(path, [
-							...warnedPackaged.get(path),
-							opts.importer,
-						]);
+					let pathWarnedPackaged = warnedPackaged.get(path);
+					if (pathWarnedPackaged === undefined) {
+						warnedPackaged.set(path, (pathWarnedPackaged = []));
 					}
+					pathWarnedPackaged.push(opts.importer);
+
 					return { external: true };
 				}
 				// This is a normal package—don't treat it specially

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts
@@ -3,12 +3,6 @@ import chalk from "chalk";
 import { logger } from "../../logger";
 import type { Plugin } from "esbuild";
 
-// Infinite loop detection
-const seen = new Set();
-
-// Prevent multiple warnings per package
-const warnedPackaged = new Map();
-
 /**
  * An esbuild plugin that will mark any `node:...` imports as external.
  */
@@ -17,8 +11,16 @@ export const nodejsCompatPlugin: (silenceWarnings: boolean) => Plugin = (
 ) => ({
 	name: "nodejs_compat imports plugin",
 	setup(pluginBuild) {
-		seen.clear();
-		warnedPackaged.clear();
+		// Infinite loop detection
+		const seen = new Set();
+
+		// Prevent multiple warnings per package
+		const warnedPackaged = new Map();
+
+		pluginBuild.onStart(() => {
+			seen.clear();
+			warnedPackaged.clear();
+		});
 		pluginBuild.onResolve(
 			{ filter: /node:.*/ },
 			async ({ path, kind, resolveDir, ...opts }) => {

--- a/packages/wrangler/src/navigator-user-agent.ts
+++ b/packages/wrangler/src/navigator-user-agent.ts
@@ -1,16 +1,16 @@
-import assert from "node:assert";
+import { UserError } from "./errors";
 
 export function isNavigatorDefined(
 	compatibility_date: string | undefined,
 	compatibility_flags: string[] = []
 ) {
-	assert(
-		!(
-			compatibility_flags.includes("global_navigator") &&
-			compatibility_flags.includes("no_global_navigator")
-		),
-		"Can't both enable and disable a flag"
-	);
+	if (
+		compatibility_flags.includes("global_navigator") &&
+		compatibility_flags.includes("no_global_navigator")
+	) {
+		throw new UserError("Can't both enable and disable a flag");
+	}
+
 	if (compatibility_flags.includes("global_navigator")) {
 		return true;
 	}

